client.go

```diff
@@ -1207,46 +1207,47 @@ func (t *Torrent) MergeSpec(spec *TorrentSpec) error {
 }
 
 func useTorrentSources(sources []string, t *Torrent) {
-	for _, s := range sources {
-		go func(s string) {
-			err := useTorrentSource(s, t)
-			if err != nil {
+	ctx := context.Background()
+	for i := 0; i < len(sources); i += 1 {
+		s := sources[i]
+		go func() {
+			if err := useTorrentSource(ctx, s, t); err != nil {
 				t.logger.WithDefaultLevel(log.Warning).Printf("using torrent source %q: %v", s, err)
 			} else {
 				t.logger.Printf("successfully used source %q", s)
 			}
-		}(s)
+		}()
 	}
 }
 
-func useTorrentSource(source string, t *Torrent) error {
-	req, err := http.NewRequest(http.MethodGet, source, nil)
-	if err != nil {
-		panic(err)
-	}
-	ctx, cancel := context.WithCancel(context.Background())
+func useTorrentSource(ctx context.Context, source string, t *Torrent) (err error) {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
 	defer cancel()
 	go func() {
 		select {
 		case <-t.GotInfo():
 		case <-t.Closed():
-		case <-ctx.Done():
+		case <-ctxWithCancel.Done():
 		}
 		cancel()
 	}()
-	req = req.WithContext(ctx)
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return err
+	var req *http.Request
+	if req, err = http.NewRequestWithContext(ctxWithCancel, http.MethodGet, source, nil); err != nil {
+		panic(err)
 	}
-	mi, err := metainfo.Load(resp.Body)
-	if err != nil {
+	var resp *http.Response
+	if resp, err = http.DefaultClient.Do(req); err == nil {
+		var mi metainfo.MetaInfo
+		err = bencode.NewDecoder(resp.Body).Decode(&mi)
```

Comment:
Why not use metainfo.Load here? I think it actually has different behaviour, (nevermind that it probably shouldn't exist).

Reply:
Tbh I just didn't like how `metainfo.Load` returned a pointer instead of a value (especially since it declares `var mi metainfo.MetaInfo` only to return `&mi` -- thus relying on escape analysis). I wanted to restrict the lifetime of `mi` as much as possible, without relying on the GC to collect it later from the heap.

But if you insist on using `metainfo.Load`, I guess we could do something like
```
var mi *metainfo.MetaInfo
mi, err = metainfo.Load(resp.Body)
```
and rely on escape analysis to determine how to clean up `mi`. IMO it's better to just scope `mi`.

Reply:
Yep, keep this the way you have, although note escape analysis will probably put it on the heap anyway.